plugins/test/framework.cc

```diff
@@ -20,13 +20,54 @@
 
 namespace service_extensions_samples {
 
+size_t Buffer::size() const {
+  if (owned_string_buffer_) {
+    return owned_string_buffer_->length();
+  }
+  return proxy_wasm::BufferBase::size();
+}
+proxy_wasm::WasmResult Buffer::copyTo(proxy_wasm::WasmBase* wasm, size_t start,
+                                      size_t length, uint64_t ptr_ptr,
+                                      uint64_t size_ptr) const {
+  if (owned_string_buffer_) {
+    auto buffer_section = owned_string_buffer_->substr(start, length);
```

Comment:
`substr()` returns a `std::string`, which means that this line incurs an extra copy in addition to the later (necessary) copy into the Wasm VM. To avoid this, do:

```
if (start + length > owned_string_buffer_.size()) {
  return WasmResult::InvalidMemoryAccess;
}
std::string_view s(owned_string_buffer_.data() + start, length);
if (!wasm->copyToPointerSize(s, ptr_ptr, size_ptr)) { ... }
```